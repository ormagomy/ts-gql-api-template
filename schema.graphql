# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

input CreateTodoInput {
  content: String!
  isDone: Boolean
}

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.This scalar is serialized to a string in ISO 8601 format and parsed from a string in ISO 8601 format.
"""
scalar DateTimeISO

type Mutation {
  createTodo(createTodoInput: CreateTodoInput!): Todo!
  updateTodo(updateTodoInput: UpdateTodoInput!): Todo!
}

"""Mongo object id scalar type"""
scalar ObjectId

type Query {
  todo(id: ObjectId!): Todo!
  todos: [Todo!]!
}

type Todo {
  _id: ObjectId!
  content: String!
  createdAt: DateTimeISO!
  isDone: Boolean!
  updatedAt: DateTimeISO!
}

input UpdateTodoInput {
  _id: ObjectId!
  content: String
  isDone: Boolean
}