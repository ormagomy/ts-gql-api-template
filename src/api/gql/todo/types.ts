import { MaxLength, MinLength, NotEquals } from 'class-validator'
import { Field, InputType, ObjectType } from 'type-graphql'

import { withId } from '@src/utils/mixins/with.id'
import { withTimestamps } from '@src/utils/mixins/with.timestamps'

@ObjectType()
@InputType()
class TodoBase {
  @Field({ nullable: true })
  @MaxLength(300)
  @MinLength(1)
  content?: string

  @Field({ nullable: true })
  @NotEquals(null)
  isDone?: boolean
}

@ObjectType()
export class Todo extends withTimestamps(withId(TodoBase)) {
  @Field()
  content: string

  @Field()
  isDone: boolean
}

@InputType()
export class CreateTodoInput extends TodoBase {
  @Field()
  content: string
}

@InputType()
export class UpdateTodoInput extends withId(TodoBase) {}
