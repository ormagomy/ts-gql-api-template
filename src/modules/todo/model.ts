import { getModelForClass } from '@typegoose/typegoose'
import { ObjectId } from 'mongodb'
import { Service } from 'typedi'

import { CreateTodoType, Todo, UpdateTodoType } from '@src/modules/todo/entity'
import { NotFoundError } from '@src/utils/errors'

export const TodoMongooseModel = getModelForClass(Todo)

@Service()
export default class TodoModel {
  async getById(_id: ObjectId): Promise<Todo | null> {
    return TodoMongooseModel.findById(_id).lean().exec()
  }

  async getAll(): Promise<Todo[]> {
    return TodoMongooseModel.find().lean().exec()
  }

  async create(data: CreateTodoType): Promise<Todo> {
    return TodoMongooseModel.create(data)
  }

  async update({ _id, ...data }: UpdateTodoType): Promise<Todo> {
    const todo = await TodoMongooseModel.findById({ _id })

    if (!todo) {
      throw new NotFoundError()
    }

    todo.set(data)
    return todo.save()
  }
}
