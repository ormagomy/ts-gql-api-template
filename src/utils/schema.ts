import { ObjectId } from 'mongodb'
import { buildSchemaSync } from 'type-graphql'
import Container from 'typedi'

import { resolvers } from '@src/api/gql'
import { ObjectIdScalar } from '@src/utils/scalars'

// Here goes your schema building bit, doing it this way allows us to use it in the tests as well!
export const buildSchema = () =>
  buildSchemaSync({
    resolvers,
    container: Container,
    scalarsMap: [{ type: ObjectId, scalar: ObjectIdScalar }],
    validate: {
      skipMissingProperties: false,
      skipUndefinedProperties: true,
      skipNullProperties: false,
    },
    emitSchemaFile: true,
  })
