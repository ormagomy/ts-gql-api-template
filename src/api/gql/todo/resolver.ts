import { ObjectId } from 'mongodb'
import { Arg, ArgsType, Field, Int, Mutation, Query, Resolver } from 'type-graphql'
import { Service } from 'typedi'

import { CreateTodoInput, Todo, UpdateTodoInput } from './types'
import TodoService from '@src/modules/todo/service'

// TODO: think about pagination
@ArgsType()
// eslint-disable-next-line @typescript-eslint/no-unused-vars
class GetAllArgs {
  @Field(() => Int, { nullable: true })
  skip?: number

  @Field(() => Int, { nullable: true })
  take?: number
}

@Service()
@Resolver(Todo)
export default class TodoResolver {
  constructor(private readonly todoService: TodoService) {}

  @Query(() => Todo)
  async todo(@Arg('id') id: ObjectId) {
    const todo = await this.todoService.getById(id)
    return todo
  }

  @Query(() => [Todo])
  async todos() {
    const todos = await this.todoService.getAll()
    return todos
  }

  @Mutation(() => Todo)
  async createTodo(@Arg('createTodoInput') createTodoInput: CreateTodoInput): Promise<Todo> {
    const todo = await this.todoService.addTodo(createTodoInput)
    return todo
  }

  @Mutation(() => Todo)
  async updateTodo(@Arg('updateTodoInput') updateTodoInput: UpdateTodoInput): Promise<Todo> {
    const todo = await this.todoService.updateTodo(updateTodoInput)
    return todo
  }
}
