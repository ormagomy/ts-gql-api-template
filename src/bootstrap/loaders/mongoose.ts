import mongoose from 'mongoose'

import { config } from '@src/config'

// Close the Mongoose default connection in the event of application termination
process.on('SIGINT', async () => {
  console.log('Closing Mongoose connection...')
  await mongoose.connection.close()
  process.exit(0)
})

const mongoDBConfig: mongoose.ConnectOptions = {
  dbName: config.mongoDB.dbName,
  connectTimeoutMS: 5000,
  serverSelectionTimeoutMS: 5000,
  socketTimeoutMS: 5000,
}

export default (): Promise<mongoose.Mongoose> => mongoose.connect(config.mongoDB.uri, mongoDBConfig)
