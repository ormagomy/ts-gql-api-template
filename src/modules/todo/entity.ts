import { modelOptions, prop } from '@typegoose/typegoose'

import { withId } from '@src/utils/mixins/with.id'
import { withTimestamps } from '@src/utils/mixins/with.timestamps'
import { partialOf, partialWithId } from '@src/utils/typeHelpers'

export class TodoBase {
  @prop()
  content: string

  @prop({ default: false })
  isDone: boolean
}

@modelOptions({
  schemaOptions: {
    collection: 'todos',
    timestamps: true,
  },
})
export class Todo extends withTimestamps(withId(TodoBase)) {}

export type CreateTodoType = partialOf<TodoBase, 'content'>
export type UpdateTodoType = partialWithId<Todo>
