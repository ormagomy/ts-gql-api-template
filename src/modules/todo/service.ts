import { ObjectId } from 'mongodb'
import { Service } from 'typedi'

import { CreateTodoType, Todo, UpdateTodoType } from '@src/modules/todo/entity'
import TodoModel from './model'

@Service()
export default class TodoService {
  constructor(private readonly todoModel: TodoModel) {}

  public async getById(_id: ObjectId): Promise<Todo | null> {
    return this.todoModel.getById(_id)
  }

  public async getAll(): Promise<Todo[]> {
    return this.todoModel.getAll()
  }

  public async addTodo(data: CreateTodoType): Promise<Todo> {
    const newTodo = await this.todoModel.create(data)

    // Business logic goes here
    // Example: Trigger push notification, analytics, ...

    return newTodo
  }

  public async updateTodo(data: UpdateTodoType): Promise<Todo> {
    const newTodo = await this.todoModel.update(data)

    // Business logic goes here
    // Example: Trigger push notification, analytics, ...

    return newTodo
  }
}
