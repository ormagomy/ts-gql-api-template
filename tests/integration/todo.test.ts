import { ApolloServer } from '@apollo/server'
import { gql } from 'graphql-tag'
import mongoose from 'mongoose'

import { TodoMongooseModel } from '@src/modules/todo/model'
import { buildSchema } from '@src/utils/schema'

import { clearDatabase, closeDatabase, connect, populateDatabase } from '@tests/utils'

beforeAll(async () => connect())

// You can populate de DB before each test
beforeEach(async () => {
  await populateDatabase(TodoMongooseModel, [
    {
      content: 'todo 1',
      isDone: false,
    },
    {
      content: 'todo 2',
      isDone: true,
    },
  ])
})

afterEach(async () => {
  await clearDatabase()
})

afterAll(async () => {
  await closeDatabase()
})

describe('Todo', () => {
  it(`should create a todo`, async () => {
    const graphQLSchema = buildSchema()

    const server = new ApolloServer({
      schema: graphQLSchema,
    })

    // Define the query and the variables as you would do from your front-end
    const variables = {
      createTodoInput: {
        content: `Test todo.`,
      },
    }

    const CREATE_TODO = gql`
      mutation createTodo($createTodoInput: CreateTodoInput!) {
        createTodo(createTodoInput: $createTodoInput) {
          content
        }
      }
    `

    // Run query against the server and snapshot the output
    const res = await server.executeOperation({
      query: CREATE_TODO,
      variables,
    })

    expect(res).toMatchSnapshot()
  })

  it(`should get a todo`, async () => {
    const todoId = new mongoose.Types.ObjectId().toHexString().toString()

    // Add a todo with the pre-generated ID in the database
    await populateDatabase(TodoMongooseModel, [
      {
        _id: todoId,
        content: 'todo 1',
        isDone: false,
      },
    ])

    const graphQLSchema = buildSchema()

    const server = new ApolloServer({
      schema: graphQLSchema,
    })

    const variables = {
      id: todoId,
    }

    const GET_TODO = gql`
      query todo($id: ObjectId!) {
        todo(id: $id) {
          content
        }
      }
    `

    const res = await server.executeOperation({
      query: GET_TODO,
      variables,
    })

    expect(res).toMatchSnapshot()
  })

  it(`should get all todos`, async () => {
    const graphQLSchema = buildSchema()

    const server = new ApolloServer({
      schema: graphQLSchema,
    })

    const GET_ALL_TODOS = gql`
      query getAllTodos {
        todos {
          content
          isDone
        }
      }
    `

    const res = await server.executeOperation({
      query: GET_ALL_TODOS,
    })

    expect(res).toMatchSnapshot()
  })

  it(`should update a todo`, async () => {
    const todoId = new mongoose.Types.ObjectId().toHexString().toString()

    // Add a todo with the pre-generated ID in the database
    await populateDatabase(TodoMongooseModel, [
      {
        _id: todoId,
        content: 'todo 1',
        isDone: false,
      },
    ])

    const graphQLSchema = buildSchema()

    const server = new ApolloServer({
      schema: graphQLSchema,
    })

    const variables = {
      updateTodoInput: {
        _id: todoId,
        content: 'Updated todo',
        isDone: true,
      },
    }

    const UPDATE_TODO = gql`
      mutation updateTodo($updateTodoInput: UpdateTodoInput!) {
        updateTodo(updateTodoInput: $updateTodoInput) {
          content
          isDone
        }
      }
    `

    const res = await server.executeOperation({
      query: UPDATE_TODO,
      variables,
    })

    expect(res).toMatchSnapshot()
  })

  it(`should return an error when updating a todo with a non-existent _id`, async () => {
    const missingId = new mongoose.Types.ObjectId().toHexString().toString()

    const graphQLSchema = buildSchema()

    const server = new ApolloServer({
      schema: graphQLSchema,
    })

    const variables = {
      updateTodoInput: {
        _id: missingId,
        content: 'Updated todo',
      },
    }

    const UPDATE_TODO = gql`
      mutation updateTodo($updateTodoInput: UpdateTodoInput!) {
        updateTodo(updateTodoInput: $updateTodoInput) {
          content
          isDone
        }
      }
    `

    const res = await server.executeOperation({
      query: UPDATE_TODO,
      variables,
    })

    expect(res).toMatchSnapshot()
  })
})
