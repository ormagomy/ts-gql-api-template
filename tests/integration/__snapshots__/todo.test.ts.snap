// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should create a todo 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "createTodo": {
          "content": "Test todo.",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Todo should get a todo 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "todo": {
          "content": "todo 1",
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Todo should get all todos 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "todos": [
          {
            "content": "todo 1",
            "isDone": false,
          },
          {
            "content": "todo 2",
            "isDone": true,
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Todo should return an error when updating a todo with a non-existent _id 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": null,
      "errors": [
        {
          "extensions": {
            "code": "INTERNAL_SERVER_ERROR",
          },
          "locations": [
            {
              "column": 3,
              "line": 2,
            },
          ],
          "message": "Entity not found",
          "path": [
            "updateTodo",
          ],
        },
      ],
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;

exports[`Todo should update a todo 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "updateTodo": {
          "content": "Updated todo",
          "isDone": true,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
